include/up-cpp/client/RpcTarget.h

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-FileType: SOURCE
+// SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef UP_CPP_CLIENT_RPCTARGET_H
+#define UP_CPP_CLIENT_RPCTARGET_H
+
+#include <uprotocol/v1/umessage.pb.h>
+#include <uprotocol/v1/uri.pb.h>
+#include <uprotocol/v1/ustatus.pb.h>
+#include <up-cpp/datamodel/builder/UMessage.h>
+#include <up-cpp/datamodel/validator/UMessage.h>
+#include <up-cpp/transport/UTransport.h>
+
+#include <memory>
+#include <utility>
+
+namespace uprotocol::client {
+
+/// @brief Interface for uEntities to act as a target of RPC requests.
+///
+/// Like all L2 client APIs, the RpcTarget is a wrapper on top of the L1
+/// UTransport API; in this instance, it is the request-handling half of the
+/// RPC model.
+struct RpcTarget {
+	/// @brief Callback function signature for implementing the RPC method.
+	using RpcMethod = transport::UTransport::ListenCallback;
+
+	/// @brief Constructs an initiator connected to a given transport.
+	///
+	/// @param transport Transport to offer the RPC method through.
+	/// @param method_name URI representing the name clients will use to invoke
+	///                    the RPC method.
+	/// @param callback Method that will be called when requests are received.
+	RpcTarget(std::shared_ptr<transport::UTransport> transport,
+	          const v1::UUri& method_name, RpcMethod&& callback);
+
+	/// @brief Builder type for messages sent by this client
+	using MessageBuilder = datamodel::builder::UMessageBuilder;
+
+	/// @brief Gets a UMessageBuilder by forwarding parameters to
+	///        UMessageBuilder::response().
+	///
+	/// @remarks Generally, passing the request message this response is being
+	///          sent for is the safest option.
+	///
+	/// @see datamodel::builder::UMessageBuilder::response
+	template <typename... Args>
+	static MessageBuilder responseBuilder(Args&&... args) {
+		return MessageBuilder::response(std::forward<Args>(args)...);
+	}
+
+	/// @brief Sends a response message.
+	///
+	/// Intended to pair with responseBuilder().
+	///
+	/// @throws InvalidUMessage if the message is not a response.
```

Comment:
This is where I'm always torn on exceptions.

We have a return of `UStatus`. In theory, if something bad happened we should be able to capture what went wrong within there, no?

Reply:
See #79 where we discussed this. The short summary is that exceptions are thrown when the programmer messed up. Statuses are returned when there is a transient failure.

"You tried to respond with a request message" is a exception.
"I tried to send your response, but it failed because the network connection was lost" is a status message.